### ISSUE TITLE
Better way to get calculated values without updating UI.

### ISSUE THREAD

**Issue by pkpatels** [OfficeDev/office-js#1826]
> Message from office-js bot:
We’re closing this issue because it has been inactive for a long time. We’re doing this to keep the issues list manageable and useful for everyone. If this issue is still relevant for you, please create a new issue.
> Thank you for your understanding and continued feedback.

We have around 10,000 cells, scattered across ranges and worksheets (not adjacent).

Looping through each of them, one by one, we have to change the value of a cell, recalculate an output cell, reset the change cell to the original value and move to the next cell.

We have already tried following things but it doesn't help.

1. We buffer the ranges and call `context.sync()` only once.
2. we also call `app.suspendScreenUpdatingUntilNextSync()` in the beginning (check line **no 21** in [gist url](https://gist.github.com/pkpatels/8b3bca619b39c78cb0bc09e18c8512c1#file-sync-yaml-L21)), so it doesn't update UI on every change/calculations. 
3. I can't use `app.suspendApiCalculationUntilNextSync()` because I want to get calculated values of a particular (single) cell by changing other cells one by one. We must have a calculated value at every single change in a cell like shown in [gist url](https://gist.github.com/pkpatels/038822c76953cb88fd6db34e558554ca).

The buffer contains more than 10,000 ranges/cells and when we call `context.sync()` I see white screen only (Excel cells are not visible). So can you please provide a better way to get **calculated values** without updating UI.

Here is the [gist url](https://gist.github.com/pkpatels/8b3bca619b39c78cb0bc09e18c8512c1) that simulate the situation I have. Please use excel file added in comment section (this is just example where cell are adjacent but in actual scenario the cells are scattered across ranges and worksheets).

From [Stackoverflow](https://stackoverflow.com/questions/66981426/get-excel-cell-value-without-ui-update-in-excel-addin-nodejs-api) we got a suggestion for feature request for "**Excel session level UI suspend**" API.

**Comment by jipinshi** [comment]
@pkpatels, thanks let us know it.  It seems there are a lot of API calls in the two for loop which caused the bad perf. Suggest to put data in a 2 dimension array values, then with range.values = values  to do similar thing, which makes it faster.

**Comment by pkpatels** [comment]
@jipinshi Thanks for the response. 

Actually we need to get calculated value of a cell while changing cells one by one.

e.g. 
formula of **A1** is **=B1+C2+D3**

Now I want to get calculated value of A1 
1 - by changing only value of B1
2 - by changing only value of C2
3 - by changing only value of D3

So in our situation I can't change values of B1, C2 and D3 in a single range as you suggest using two dimensional array. and get calculated value. 

I hope I make my situation clear to you. Let me know If you need more detail.

**Comment by jipinshi** [comment]
@pkpatels, hope range.formulas and ctx.workbook.application.suspendApiCalculationUntilNextSync() can help here.

check [here]( https://docs.microsoft.com/en-us/office/dev/add-ins/excel/performance )

**Comment by pkpatels** [comment]
@jipinshi, I appreciate your suggestion but I already tried that and I didn't get the desired result. 

Because here I want to get calculated values of a particular (single) cell by changing other cells one by one. In this case if we **suspendApiCalculationUntilNextSync** it calculate the value based on last change in cells (ignore the previous changes in the cell)

Here is the [gist url](https://gist.github.com/pkpatels/038822c76953cb88fd6db34e558554ca) that simulate the situation I have.

**Comment by gingerjia** [comment]
@pkpatels, seems your requirement is to suspend UI update while calculate values to better the perf. If this purpose, then I think 'suspendScreenUpdatingUntilNextSync' can meet your requirement. Here is the details: https://docs.microsoft.com/en-us/office/dev/add-ins/excel/performance 

And for your gist https://gist.github.com/pkpatels/8b3bca619b39c78cb0bc09e18c8512c1, there are couple of areas can be optimized to improve the performance also. For example:
    for (let r = 0; r < range.values.length; r++) {
          let row = range.rowIndex + r;
          for (let c = 0; c < range.values[r].length; c++) {
            let col = range.columnIndex + c;
The range.values.length, range.rowIndex and range.columnIndex should be a const value, you can define var to store them at first, then reference at the for loop:
    var rowIndex = range.rowIndex;
    var columnIndex = range.columnIndex;
    var valuesLength = range.values.length;

    for (let r = 0; r < valuesLength; r++) {
      let row = rowIndex + r;
      for (let c = 0; c < range.values[r].length; c++) {
        let col = columnIndex + c;

And here you can get more performance related best practice: https://docs.microsoft.com/en-us/office/dev/add-ins/concepts/resource-limits-and-performance-optimization#excel-add-ins
Thanks!

**Comment by pkpatels** [comment]
Thank you @gingerjia, as I mentioned above we already use `app.suspendScreenUpdatingUntilNextSync()` (https://github.com/OfficeDev/office-js/issues/1826#issue-868584238 check line no 21 in [gist url](https://gist.github.com/pkpatels/8b3bca619b39c78cb0bc09e18c8512c1#file-sync-yaml-L21))  and I can't use  `app.suspendApiCalculationUntilNextSync()` because https://github.com/OfficeDev/office-js/issues/1826#issuecomment-829804680.  Unfortunately a screen update is triggered anyway because of `context.sync()` and I see white screen only (Excel cells are not visible).

Can you suggest another approach?

**Comment by gingerjia** [comment]
Currently we only has context.sync level suspend UI update API. We have logged work item for your new requirement as session level based suspend UI update. 
For your side issue, can you divide the more than 10,000 ranges/cells buffer update into multiple context.sync()? And can you optimize your script according our best practice https://docs.microsoft.com/en-us/office/dev/add-ins/concepts/resource-limits-and-performance-optimization#excel-add-ins?

**Comment by pkpatels** [comment]
@gingerjia yes. I already update buffer into multiple (100) `context.sync()`. It reduce the time for a single `context.sync()` but still I see white screen only (Excel cells are not visible) for some time on every `context.sync()`

**Comment by jipinshi** [comment]
@pkpatels, can you use autofill to replace  sheet1.getCell(row, col).formulas = [[Math.random()]]; for each cell? Refer to [this](https://docs.microsoft.com/en-us/javascript/api/excel/excel.range?view=excel-js-preview#autoFill_destinationRange__autoFillType_) and [this](https://docs.microsoft.com/en-us/javascript/api/excel/excel.range?view=excel-js-preview#autoFill_destinationRange__autoFillTypeString_), each time to call one api to set formula it will make it slow and blank.

**Comment by pkpatels** [comment]
@jipinshi Thank you for your time. To elaborate more, the cells are scattered across ranges and worksheets (not adjacent). So I don't think autofill can help here. 

To give you more idea I update the description (original text). Kindly check it once.

**Comment by pkpatels** [comment]
@jipinshi I thought about one solution to avoid `context.sync()` call. 

The idea is that I can subscribe to **WorksheetCalculatedEvent**. But **WorksheetCalculatedEvent** do not pass information about calculated value ([Documentation](https://docs.microsoft.com/en-us/javascript/api/excel/excel.worksheetcalculatedeventargs?view=excel-js-preview)). 

Can you provide the [details](https://docs.microsoft.com/en-us/javascript/api/excel/excel.changedeventdetail?view=excel-js-preview#excel_Excel_ChangedEventDetail_valueBefore_member) property in the **WorksheetCalculatedEvent**, like [WorksheetChangedEvent](https://docs.microsoft.com/en-us/javascript/api/excel/excel.worksheetchangedeventargs?view=excel-js-preview#details)?

**Comment by lumine2008** [comment]
@pkpatels ， would you kindly share your scenario, as WorksheetCalculatedEvent event happens frequently, more information provided in this event may result in the perf impact. therefore we would like to learn your scenario and see if there is any solution to unblock your scenario.  Thanks.

**Comment by pkpatels** [comment]
@lumine2008 Thanks for the response. 

I tried to describe the scenario in the description [here](https://github.com/OfficeDev/office-js/issues/1826#issue-868584238). Is it clear or do you require any specific additional information?

**Comment by lumine2008** [comment]
I would suggest discussing via mail. do you mind to send me a mail: raymonlu@microsoft.com, 
if needed, we could schedule a meeting to discuss in detail. thanks.
